storage, compute: copy to s3 with multiple workers

moulimukherjee opened the pull request:
<!--
Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.
-->


### Motivation
Changes to write to s3 from multiple workers, part of https://github.com/MaterializeInc/materialize/issues/7256

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer
Opening this PR to get some early feedback. I will try to get the changes to make the bucket size configurable by tomorrow.

@benesch If I am not able to get this merged by this week, somebody else will have to pick this up.

cc @petrosagg @guswynn 
<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

moulimukherjee commented on src/compute/src/sink/copy_to_s3_oneshot.rs:
```diff
@@ -58,22 +57,29 @@ where
+        // we can write files to s3 deterministically across different replicas of different sizes
+        // using the bucket ID. Each worker will split a bucket's data into 1 or more
+        // files based on the user provided `MAX_FILE_SIZE`.
+        // TODO(#7256): make the number of buckets configurable.
+        let bucket_count = 20;
```
I will try to get this in by tomorrow in this PR. But opening the PR now to get early feedback.

moulimukherjee replied:
done

moulimukherjee replied:
Will try to get the max number of workers across replicas instead
https://github.com/MaterializeInc/materialize/commit/4d0fd27e976ad851bf91c5a89bb7a245510eb55a#r139476753

moulimukherjee commented on src/compute/src/sink/copy_to_s3_oneshot.rs:
```diff
@@ -58,22 +57,29 @@ where
+
         let input = consolidate_pact::<KeyBatcher<_, _, _>, _, _, _, _, _>(
-            &sinked_collection.map(|row| (row, ())),
-            exchange,
+            &sinked_collection.map(move |row| {
+                let bucket = row.hashed() % bucket_count;
+                ((row, bucket), ())
+            }),
+            Exchange::new(move |(((_, bucket), _), _, _)| *bucket),
             "Consolidated COPY TO S3 input",
         );
```
We will need the data to be in a deterministic order after the exchange. Is that a correct assumption to make?

moulimukherjee replied:
Is there any way to actually get the data in an order? The splitting into multiple files will not work otherwise.

benesch replied:
I think that's fine though, right? The property we need is that two replicas with different worker counts that are executing the same `COPY TO S3` operation need to put the same rows in the same CSV files. I don't think the order of the rows in those CSV files matters, at long as the multiset of rows is identical.

moulimukherjee replied:
@benesch But it will be a problem if a worker has to split the data into multiple files to stay under MAX_FILE_SIZE? The workers in different replicas will have different data in the splits then.

moulimukherjee replied:
er...should we just ignore the max_file_size for the time being and set like a large value internally like 5GB?

petrosagg replied:
The data will come out of this operator in a deterministic order because here we are defining the pact of consolidation which will sort the data. We don't really care that the exchange pact maintains the order (even though it does). @antiguru does that check out?

antiguru replied:
Yes, but it's not a property we support -- it might work right now, but it's relying on something that's not documented as such. I'm fine with leaving it as-is for the moment if there's a follow-up issue filed: Instead of using `consolidate_pact`, the implementation could hook into the merge batcher (as `consolidate_pact` does under the hood) and observe the correctly-ordered stream of updates, without transitioning through a stream.

moulimukherjee replied:
Ack, ty. Filed https://github.com/MaterializeInc/materialize/issues/25835 to track and will also add a comment in the code to that effect.

moulimukherjee commented on src/storage-operators/src/s3_oneshot_sink.rs:
```diff
@@ -123,17 +122,18 @@ pub fn copy_to<G, F>(
-                            Err(e) => {
-                                onetime_callback(Err(e.to_string()));
-                                return;
+                        for uploader in s3_uploaders.values_mut() {
+                            match uploader.flush().await {
+                                Ok(()) => {}
+                                Err(e) => {
+                                    onetime_callback(Err(e.to_string()));
+                                    return;
+                                }
                             }
                         }
+                        // We are done, send the final count.
+                        onetime_callback(Ok(row_count));
+                        return;
```
Yes, I think that should be fine. Anytime we stop the upload, either due to an error or user cancellation, it should be expected that some data was already uploaded.

jkosh44 reviewed (approved):
Adapter changes LGTM

moulimukherjee commented on src/controller/src/clusters.rs:
```diff
@@ -197,6 +197,16 @@ impl ReplicaLocation {
+            }
+            ReplicaLocation::Unmanaged(UnmanagedReplicaLocation { workers, .. }) => *workers,
+        };
+        workers_per_process * self.num_processes()
```
(I feel like I should know this 🙈) Is this the correct way to get the number of workers for a replica?

benesch replied:
Looks right to me!

moulimukherjee replied:
ty!

moulimukherjee commented on src/adapter/src/coord/sequencer/inner/peek.rs:
```diff
@@ -315,7 +324,17 @@ impl Coordinator {
+                    .map(|r| r.config.location.workers())
+                    .max()
+                    .map(u64::cast_from);
+                assert!(max_worker_count.is_some());
```
This should always be `Some` because the check to see if there are available replicas now happen before this.

benesch reviewed (commented):
Only skimmed the max worker count logic on mobile (sorry, on a plane) but the shape of it all LGTM!

moulimukherjee commented on src/adapter/src/coord/sequencer/inner/peek.rs:
```diff
@@ -315,7 +324,17 @@ impl Coordinator {
+                // Getting the max worker count across replicas
+                // and using that value for the number of buckets to
+                // divide the copy to output into.
+                let max_worker_count = cluster
```
cc @benesch Updated to use the max worker count across replicas as the number of buckets.

moulimukherjee commented:
👋 Have addressed all the comments, and should be good for another round of review. 

cc @jkosh44 I updated the adapter bits again based on feedback 🙈, can you please take another look?

jkosh44 reviewed (approved):
Adapter code LGTM

def- reviewed (commented):
Sorry I didn't have time for the proper testing yet.

Some interesting uncovered lines marked inline based on https://buildkite.com/materialize/coverage/builds/386

moulimukherjee commented on src/storage-operators/src/s3_oneshot_sink.rs:
```diff
@@ -109,6 +101,13 @@ pub fn copy_to<G, F>(
                                 return;
                             }
                             row_count += u64::try_from(diff).unwrap();
+                            let uploader = s3_uploaders
+                                .entry(bucket)
+                                .or_insert_with(|| {
+                                    info!("worker_id: {} will be handling bucket: {}", worker_id, bucket);
+                                    let file_name_prefix = format!("part-bucket-{:04}", bucket);
+                                    CopyToS3Uploader::new(sdk_config.clone(), connection_details.clone(), file_name_prefix)
+                                });
```
@def- These are all actually covered by the aws-localstack tests in nightlies

moulimukherjee commented on src/storage-operators/src/s3_oneshot_sink.rs:
```diff
@@ -123,17 +122,18 @@ pub fn copy_to<G, F>(
-                                onetime_callback(Err(e.to_string()));
-                                return;
+                        for uploader in s3_uploaders.values_mut() {
+                            match uploader.flush().await {
+                                Ok(()) => {}
+                                Err(e) => {
+                                    onetime_callback(Err(e.to_string()));
+                                    return;
+                                }
```
Same ^, covered in aws-localstack nightly test

petrosagg commented on src/adapter/src/coord/sequencer/inner/peek.rs:
```diff
@@ -317,7 +326,17 @@ impl Coordinator {
+                    .map(|r| r.config.location.workers())
+                    .max()
+                    .map(u64::cast_from);
+                assert!(max_worker_count.is_some());
```
I'd feel better if this computation was incorporated in the check above such that they never get out of sync
```suggestion
        let max_worker count = match cluster.replicas().map(|r| r.config.location.workers()).max() {
            Some(count) => u64::cast_from(count),
            None => {
                return Err(AdapterError::NoClusterReplicasAvailable(
                    cluster.name.clone(),
                ));
            }
        }
```

petrosagg commented on src/testdrive/src/action/s3.rs:
```diff
@@ -16,27 +16,43 @@ pub async fn run_verify_data(
     state: &State,
 ) -> Result<ControlFlow, anyhow::Error> {
-    let expected_body = cmd.input.join("\n") + "\n";
+    let expected_body = cmd.input.sort();
```
There is precedent in testdrive that if sorting is needed to validate results it is an explicit flag. Check the `sort-messages=true` flag for `kafka-verify`. Should we do the same here?

Having the ability to verify data without sorting will allow us to write a testdrive that does `COPY TO S3` with a single replica cluster and then verify that we write the data in the correct order in the file

rjobanp replied:
@nrainer-materialize @def- what do you think?

def- replied:
I agree with Petros.

rjobanp commented:
I think all the tests are good and comments addressed now! Would appreciate a final review.

Also have code ready to address the consolidation issue and will open that as a separate PR once this merges.

jkosh44 reviewed (approved):
Adapter code still LGTM (haven't looked at the rest)

guswynn commented on src/adapter/src/coord.rs:
```diff
@@ -433,6 +433,9 @@ pub struct CopyToContext {
     pub format_params: CopyFormatParams<'static>,
     /// Approximate max file size of each uploaded file.
     pub max_file_size: u64,
+    /// Number of batches the output of the COPY TO will be partitioned into
+    /// to distribute the load across workers deterministically.
+    pub output_batch_count: Option<u64>,
```
We should describe why this is an option, or, because the reason is kindof esoteric (its set later on `peek_stage_validate`), we might just want to default it to 0

guswynn commented on src/storage-operators/src/s3_oneshot_sink.rs:
```diff
@@ -109,6 +101,13 @@ pub fn copy_to<G, F>(
+                            let uploader = s3_uploaders
+                                .entry(batch)
+                                .or_insert_with(|| {
+                                    info!("worker_id: {} will be handling batch: {}", worker_id, batch);
```
probably `debug!` is better here

guswynn commented on src/storage-operators/src/s3_oneshot_sink.rs:
```diff
@@ -34,14 +35,13 @@ use timely::PartialOrder;
     aws_connection: AwsConnection,
     connection_id: GlobalId,
-    active_worker: usize,
     onetime_callback: F,
```
can we document what this is? The other fields are pretty self-evident!

guswynn reviewed (approved):
some nits but looks good!
